### ISSUE TITLE
Support full fidelity round-tripping of Pulumi.[yaml|json]

### ISSUE THREAD

**Issue by ellismg** [pulumi/pulumi#423]
When writing information to the `Pulumi.yaml` file, we end up serializing back the entire structure and that serialization works by serializing each field in the `pack.Package` in the order it was defined in the go object.

In the case of `YAML` any comments would also be removed from the file.

This means if we had a Pulumi.yaml file that looked like this:

```
runtime: nodejs
name: hello-world
```
If we read and save it it would look like this:

```
name: hello-world
runtime: nodejs
```

Ideally, we would have parsers that would turn these files not into raw go objects but some DOM that we could round trip with full fidelity.

**Comment by justinvp** [comment]
Applies to `Pulumi.<stack>.yaml` as well.

**Comment by lblackstone** [comment]
I've looked into this and can't find a YAML parser that supports full-fidelity round-tripping. It appears that the `goyaml/v3` package comes closest, but their docs include the following note:

> It's worth noting that although Node offers access into details such as line numbers, colums, and comments, the content when re-encoded will not have its original textual representation preserved. An effort is made to render the data plesantly, and to preserve comments near the data they describe, though.

With this in mind, we should at least be able to preserve the comments in the config files, but things may still move around a bit. This would also require upgrading from v2 to v3 of this package.

**Comment by lblackstone** [comment]
Have made good progress on this issue since my last update. After some more experimentation, we determined that the `github.com/goccy/go-yaml` package is suitable for this task with some additional tweaks. We forked that library to `github.com/pulumi/go-yaml`, and made the required changes to support full-fidelity round-tripping (preserving comments, spacing, etc.).

I have a proof of concept project working locally that can unmarshal a Pulumi config, add (or delete) a config value, and then marshal it back to YAML.

![yamlRoundtrip](https://user-images.githubusercontent.com/9253463/110514992-1ee26c00-80c5-11eb-8e64-59f1f9ff9b96.gif)

The next steps are to update the YAML unmarshaling/marshaling code in the pulumi CLI to use this package.

**Comment by lblackstone** [comment]
Quick update on status of this issue:

This is currently marked as "blocked" because some unforeseen compatibility issues with the new YAML parser library required us to revert the PR to add round-tripping support. The work to fix these issues was preempted by some other engineering priorities, so it's constrained by engineering investment rather than a specific issue.

The actual round-tripping work is basically done, but we need to do some more work to make sure that the YAML marshaling/unmarshaling is compatible with the existing library. Once that is done, we need to swap out the libraries again and then merge the [in-progress PR](https://github.com/pulumi/pulumi/pull/6489).

**Comment by atrauzzi** [comment]
So, I'm gonna throw a radical idea out there.  But has Pulumi considered simply *not* touching configuration files unless explicitly asked?

I talk a little bit about this odd behaviour in this ticket: https://github.com/pulumi/pulumi/issues/8402#issuecomment-966688463

Pulumi seems to expect users to manage the config files, but also seems to want to treat it as an extension of its internally managed state store.  Is there any way Pulumi can just be a _reader_ of files in source control and nothing else?

**Comment by AaronFriel** [comment]
Comment-preserving edits will be implemented in the next release of Pulumi.

**Comment by AaronFriel** [comment]
@atrauzzi Our users find `pulumi config --set` and assorted commands to be very helpful, as a scriptable interface to managing this aspect of Pulumi. I'm pleased to say that the next release will preserve comments when using commands that edit project and config files.

When editing a file like so:

```yaml
encryptionsalt: v1:ThS5UPxP9qc=:v1:UZYAXF+ylaJ0rGhv:9OTvBnOEDFgxs7btjzSu+LZ470vLpg==
# 🔴 header comment
config:
  foo:a: some-value # 🟠 comment after value
  # 🟡 comment before key
  foo:b: some-value
  foo:c:
    a: A
    b: B
    c:
      - 1
      - 2
      - 3 # 🟢 comment in array
      # 🔵 comment after array
  foo:d:
    secure: v1:T1ftqhY0hqr+EJK6:+jvd5PMecFx80tcavzuZY4tLatgIfoe/xR72GA== # 🟣 comment on secret

# 🟥 footer comment
```

After running `pulumi config set e E`, the file will contain:

```yaml
encryptionsalt: v1:ThS5UPxP9qc=:v1:UZYAXF+ylaJ0rGhv:9OTvBnOEDFgxs7btjzSu+LZ470vLpg==
# 🔴 header comment
config:
  foo:a: some-value # 🟠 comment after value
  # 🟡 comment before key
  foo:b: some-value
  foo:c:
    a: A
    b: B
    c:
      - 1
      - 2
      - 3 # 🟢 comment in array
      # 🔵 comment after array
  foo:d:
    secure: v1:T1ftqhY0hqr+EJK6:+jvd5PMecFx80tcavzuZY4tLatgIfoe/xR72GA== # 🟣 comment on secret
  foo:e: E

# 🟥 footer comment
```

**Comment by jamest-pin** [comment]
This issue conversation seems to read that anchors are supported in Pulumi.yaml

But, I tried using them:
```yaml
# Pulumi.yaml

name: my-app
runtime: yaml
description: my app

# template
tags: &defaultTags
  Environment: ${pulumi.stack}
  CreatedBy: pulumi

resources:
  ####################################
  # L O G   G R O U P S
  # --------------------------------
  # https://www.pulumi.com/registry/packages/aws/api-  docs/cloudwatch/loggroup/
  ####################################

  # template for log groups
  logGroupTemplate: &logGroupTemplate
    type: aws:cloudwatch:LogGroup
    properties:
      tags:
        <<: *defaultTags
      logGroupClass: STANDARD
      retentionInDays: 731
      skipDestroy: True # leave logs there in case we want to go back to them. They'll fall off after the retention period

  # log group for lambda admin auth
  lg_lambda_admin_auth:
    <<: *logGroupTemplate
```

and am getting
```
error: failed to discover plugin requirements: 
-error: Pulumi.yaml:1086,13-24: alias nodes are not supported; 
-error: Pulumi.yaml:1093,9-25: alias nodes are not supported; 
```

**Comment by AaronFriel** [comment]
@jamest-pin Sorry for the delay in replying! This issue was closed and often closed issues don't get a lot of attention. You may want to open a new issue - it looks like this would belong on github.com/pulumi/pulumi-yaml.
